python/cudf/cudf/core/frame.py

```diff
@@ -1044,12 +1044,29 @@ def searchsorted(
         1-D cupy array of insertion points
         """
         # Call libcudf++ search_sorted primitive
+
+        from cudf.utils.dtypes import is_scalar
+
+        scalar_flag = None
+        if is_scalar(values):
+            scalar_flag = True
+
+        if not isinstance(values, Frame):
+            values = as_column(values)
+            if values.dtype != self.dtype:
+                self = self.astype(values.dtype)
+            values = values.as_frame()
         outcol = libcudfxx.search.search_sorted(
             self, values, side, ascending=ascending, na_position=na_position
         )
 
-        # Retrun result as cupy array
-        return cupy.asarray(outcol.data_array_view)
+        # Retrun result as cupy array if the values is non-scalar
+        # If values is scalar, result is expected to be scalar.
+        result = cupy.asarray(outcol.data_array_view)
+        if scalar_flag:
+            return result[0].tolist()
```

Comment:
In what situation do we have to return a list? And isn't `result[0]` a scalar?

Reply:
We need to return a scalar value if `values` is scalar and a list when `values` is a list.
`result[0]` is not a scalar but instead an array in cupy
```python
>>> import cupy
>>> s = cupy.array([1,2,3])
>>> s[0]
array(1)
>>> s[0].tolist()
1
```

Reply:
Okay so this call is specifically to get a scalar from a cupy array and we never actually return a Python list.

Calling `.tolist()` to accomplish this reads very ambiguously. If you do `cupy.asnumpy(result[0])` does it return a scalar or a numpy 0d-array? Either way taking a path of using `cupy.asnumpy` to handle this feels much cleaner.

Reply:
`cupy.asnumpy` also seems to have the same behavior, But we can do `.item()`
```python
>>> import cupy
>>> import numpy as np
>>> s = cupy.array([1,2,3])
>>> cupy.asnumpy(s[0])
array(1)
>>> cupy.asnumpy(s[0]).item()
1
>>> s[0]
array(1)
>>> s[0].item()
1
```

Reply:
That looks infinitely better